dimod/reference/samplers/exact_solver.py

```diff
@@ -228,4 +303,59 @@ def _all_cases_dqm(dqm):
     cases = [range(dqm.num_cases(v)) for v in dqm.variables] 
     combinations = np.array(np.meshgrid(*cases)).T.reshape(-1,dqm.num_variables())
     
-    return combinations
\ No newline at end of file
+    return combinations
+    
+    
+def _all_cases_cqm(cqm):
+    """Get a numpy array containing all possible samples as lists of valid values"""
+    # developer note: the following is a catch all method, and it may be possible
+    # to streamline or refactor. It is fine for its intended use
+    
+    var_list = list(cqm.variables)
+    
+    # Set aside logical discrete variables, and the Binary variables which make them up
+    d_cases = [len(cqm.constraints[d].lhs) for d in cqm.discrete]
+    d_vars = [x for l in [[v for v in cqm.constraints[d].lhs.variables] for d in cqm.discrete] for x in l]
+    [var_list.remove(d) for d in d_vars]
```

Comment:
Ah... of course.

`d_vars` are guaranteed to not have overlapping variables because they are only pulled from `cqm.discrete`. The `cqm.add_discrete` method checks for collisions in the one-hot constraints that enforce logical discrete variables